### ISSUE TITLE
ICommand status?

### ISSUE THREAD

**Issue by MihaMarkic** [yuvaltz/Granular#8]
What't the ICommand status? Seems like a must to enable buttons.

**Comment by yuvaltz** [comment]
ICommand is not implemented, but it could be nice to support it

**Comment by MihaMarkic** [comment]
It might be tricky since in net 4 it lives in presentationcore and in 4.5+ in system, I think. I have to check it.

**Comment by MihaMarkic** [comment]
Probably the way to do it is to upgrade to .net 4.5 and add ICommand interface to Granular.Presentation.Web. I did the former, added Command property to ButtonBase, added call to ICommand.Execute on ButtonBase.OnClick and it works.
But didn't do the Web part yet.
What do you think? Should I create a pull request?

**Comment by yuvaltz** [comment]
Sure, I'll have a look are your code. Have you investigated how ICommand is supported in the WPF's sources?, just to make sure there is no requirement that you've missed (I think that commands are also connected to the buttons' IsEnabled property)
Download the sources from here http://referencesource.microsoft.com

**Comment by MihaMarkic** [comment]
Yep, it dictates IsEnabled as well through ICommand.CanExecuteChanged event. No, I didn't yet.

**Comment by diab0l** [comment]
Don't dorget about [CommandManager](https://msdn.microsoft.com/en-us/library/system.windows.input.commandmanager(v=vs.110).aspx) ([reference source](http://referencesource.microsoft.com/#PresentationCore/Core/CSharp/System/Windows/Input/Command/CommandManager.cs)) which automates the CanExecuteChanged machinery for commands. A usual pattern is to implement CanExecuteChanged like this:

    public event EventHandler CanExecuteChanged {
        add { CommandManager.RequerySuggested += value; }
        remove { CommandManager.RequerySuggested -= value; }
    }

While not strictly necessary, it would help portability of existing WPF code and frees command implementers from the burden of firing CanExecuteChanged (which resolves some tricky situations).

**Comment by MihaMarkic** [comment]
Yes, but the CommandManager is some sort of obsolete. No newer Xaml technology has it because it works too magically and is too chatty. Perhaps we could omit it?

**Comment by yuvaltz** [comment]
I don't think we can omit it completely, it'll break existing applications. We can implement a minimal version of it, but what we implement should be as compatible as possible, so we could add missing features in the future.

**Comment by MihaMarkic** [comment]
Created simple ICommand implementation, upgraded to .net 4.5 and added Moq package for unit testing. yuvaltz/Granular#15
